AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp

```diff
@@ -0,0 +1,292 @@
+  } else {
+    dev_pillar_feature_mask[ind] = 0.0f;
+  }
+}
+
+PreProcess::PreProcess(const int max_num_pillars, const int max_points_per_pillar, const int grid_x_size,
+                       const int grid_y_size, const int grid_z_size, const float pillar_x_size,
+                       const float pillar_y_size, const float pillar_z_size, const float min_x_range,
+                       const float min_y_range, const float min_z_range)
+    : max_num_pillars_(max_num_pillars),
+      max_num_points_per_pillar_(max_points_per_pillar),
+      grid_x_size_(grid_x_size),
+      grid_y_size_(grid_y_size),
+      grid_z_size_(grid_z_size),
+      pillar_x_size_(pillar_x_size),
+      pillar_y_size_(pillar_y_size),
+      pillar_z_size_(pillar_z_size),
+      min_x_range_(min_x_range),
+      min_y_range_(min_y_range),
+      min_z_range_(min_z_range) {
+  sycl::queue queue = devicemanager::GetCurrentQueue();
+
+  // allocate memory
+  dev_pillar_x_in_coors_ = sycl::malloc_device<float>(grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_, queue);
+  dev_pillar_y_in_coors_ = sycl::malloc_device<float>(grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_, queue);
+  dev_pillar_z_in_coors_ = sycl::malloc_device<float>(grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_, queue);
+  dev_pillar_i_in_coors_ = sycl::malloc_device<float>(grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_, queue);
+  dev_pillar_count_histo_ = sycl::malloc_device<int>(grid_y_size_ * grid_x_size_, queue);
+  dev_counter_ = sycl::malloc_device<int>(1, queue);
+  dev_pillar_count_ = sycl::malloc_device<int>(1, queue);
+  dev_x_coors_for_sub_ = sycl::malloc_device<float>(max_num_pillars_, queue);
+  dev_y_coors_for_sub_ = sycl::malloc_device<float>(max_num_pillars_, queue);
+}
+
+PreProcess::~PreProcess() {
+  sycl::queue queue = devicemanager::GetCurrentQueue();
+  sycl::free(dev_pillar_x_in_coors_, queue);
+  sycl::free(dev_pillar_y_in_coors_, queue);
+  sycl::free(dev_pillar_z_in_coors_, queue);
+  sycl::free(dev_pillar_i_in_coors_, queue);
+  sycl::free(dev_pillar_count_histo_, queue);
+  sycl::free(dev_counter_, queue);
+  sycl::free(dev_pillar_count_, queue);
+  sycl::free(dev_x_coors_for_sub_, queue);
+  sycl::free(dev_y_coors_for_sub_, queue);
+}
+
+void PreProcess::DoPreProcess(const float *dev_points, const int in_num_points, int *dev_x_coors, int *dev_y_coors,
+                              float *dev_num_points_per_pillar, float *dev_pillar_x, float *dev_pillar_y,
+                              float *dev_pillar_z, float *dev_pillar_i, float *dev_x_coors_for_sub_shaped,
+                              float *dev_y_coors_for_sub_shaped, float *dev_pillar_feature_mask,
+                              int *dev_sparse_pillar_map, int *host_pillar_count) {
+  sycl::queue queue = devicemanager::GetCurrentQueue();
+  queue.memset(dev_pillar_x_in_coors_, 0, grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_ * sizeof(float))
```

Comment:
Why do you .wait() after every operation   There is no dependence between then so you should be able to launch them all and then only wait at the end for all off them to complete.

Reply:
Good point. Updated that, to wait only after all memset/memcpy operations. Similar updates were made in all files.